Audio/AudioEngine.cpp

```diff
@@ -1467,111 +1474,20 @@ X3DAUDIO_HANDLE& AudioEngine::Get3DHandle() const noexcept
-    ThrowIfFailed(hr);
-
-    ComPtr<IMMDeviceCollection> devices;
-    hr = devEnum->EnumAudioEndpoints(eRender, DEVICE_STATE_ACTIVE, &devices);
-    ThrowIfFailed(hr);
-
-    ComPtr<IMMDevice> endpoint;
-    ThrowIfFailed(devices->Item(0, endpoint.GetAddressOf()));
-
-    LPWSTR id = nullptr;
-    ThrowIfFailed(endpoint->GetId(&id));
-
-    RendererDetail device;
-    device.deviceId = id;
-    device.description = L"Default";
-
-    CoTaskMemFree(id);
-
-    list.emplace_back(device);
-
-#elif defined(_XBOX_ONE)
-
-    using namespace Microsoft::WRL;
-    using namespace Microsoft::WRL::Wrappers;
-    using namespace ABI::Windows::Foundation;
-    using namespace ABI::Windows::Media::Devices;
-
-    ComPtr<IMediaDeviceStatics> mdStatics;
-    HRESULT hr = GetActivationFactory(HStringReference(RuntimeClass_Windows_Media_Devices_MediaDevice).Get(), &mdStatics);
-    ThrowIfFailed(hr);
-
-    HString id;
-    hr = mdStatics->GetDefaultAudioRenderId(AudioDeviceRole_Default, id.GetAddressOf());
-    ThrowIfFailed(hr);
-
-    RendererDetail device;
-    device.deviceId = id.GetRawBuffer(nullptr);
-    device.description = L"Default";
-    list.emplace_back(device);
-
-#elif defined(USING_XAUDIO2_REDIST) || defined(_GAMING_DESKTOP)
-
-#ifdef __MINGW32__
-    // Value matches Windows SDK header shared\devpkey.h
-    constexpr static PROPERTYKEY PKEY_Device_FriendlyName = { { 0xa45c254e, 0xdf1c, 0x4efd, { 0x80, 0x20, 0x67, 0xd1, 0x46, 0xa8, 0x50, 0xe0 } }, 14 };
-#endif
-
-    ComPtr<IMMDeviceEnumerator> devEnum;
-    HRESULT hr = CoCreateInstance(__uuidof(MMDeviceEnumerator), nullptr, CLSCTX_INPROC_SERVER, IID_PPV_ARGS(devEnum.GetAddressOf()));
-    ThrowIfFailed(hr);
-
-    ComPtr<IMMDeviceCollection> devices;
-    hr = devEnum->EnumAudioEndpoints(eRender, DEVICE_STATE_ACTIVE, &devices);
-    ThrowIfFailed(hr);
-
-    UINT count = 0;
-    ThrowIfFailed(devices->GetCount(&count));
-
-    if (!count)
-        return list;
-
-    for (UINT j = 0; j < count; ++j)
+    uint16_t GetDeviceBitDepth(const wchar_t*)
     {
-        ComPtr<IMMDevice> endpoint;
-        hr = devices->Item(j, endpoint.GetAddressOf());
-        ThrowIfFailed(hr);
-
-        LPWSTR id = nullptr;
-        ThrowIfFailed(endpoint->GetId(&id));
-
-        RendererDetail device;
-        device.deviceId = id;
-        CoTaskMemFree(id);
-
-        ComPtr<IPropertyStore> props;
-        if (SUCCEEDED(endpoint->OpenPropertyStore(STGM_READ, props.GetAddressOf())))
-        {
-            PROPVARIANT var;
-            PropVariantInit(&var);
-
-            if (SUCCEEDED(props->GetValue(PKEY_Device_FriendlyName, &var)))
-            {
-                if (var.vt == VT_LPWSTR)
-                {
-                    device.description = var.pwszVal;
-                }
-                PropVariantClear(&var);
-            }
-        }
-
-        list.emplace_back(device);
+        return 16; // This is a guess as there's no way to get this data via the Windows Runtime APIs
```

Comment:
I'm trying to see if there's a way I can get the correct bit-depth out of the Windows Runtime APIs. It's looking like it's possible, but of course a huge PITA.

Reply:
I have a solution that works using WRL for the UWP scenario. For some reason ``CreateFromIdAsyncAdditionalProperties`` fails with ``E_INVALIDARG`` when being called for the XAudio 2.8 scenario for Win32 desktop.

Reply:
The problem was that my helper classes for providing the ``IIterable`` list of additional properties when running in a Win32 desktop process required I use "FullTrust" instead of "BaseTrust" as I did wit UWP.

> The hand-written helper classes are because there's no easy to find collections for WRL with raw ABI usage. The existing collections helpers for C++ require either C++/CX (``/ZW``) or C++/WinRT projections.